src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/AttackSystem.java

```diff
@@ -22,66 +22,108 @@
         if (attackingPlayer.getComponent(CharacterHeldItemComponent.class).selectedItem.hasComponent(RaycastOnActivateComponent.class)) {
-
             // If raycast hits another player and another player has flag, make player drop flag
-            if (entity.hasComponent(PlayerCharacterComponent.class) && entity.hasComponent(CharacterHeldItemComponent.class)) {
-                CharacterHeldItemComponent characterHeldItemComponent = entity.getComponent(CharacterHeldItemComponent.class);
-                EntityRef heldItem = characterHeldItemComponent.selectedItem;
-                if (heldItem.getComponent(BlockItemComponent.class).blockFamily.getURI().toString().equals(LASUtils.BLACK_FLAG_URI)
-                        || heldItem.getComponent(BlockItemComponent.class).blockFamily.getURI().toString().equals(LASUtils.RED_FLAG_URI)) {
-                    Vector3f position = new Vector3f(attackingPlayer.getComponent(LocationComponent.class).getLocalPosition());
-                    Vector3f direction = localPlayer.getViewDirection();
-                    Vector3f newPosition = new Vector3f(position.x + direction.x * NEW_DIRECTION_MODIFIER,
-                            position.y + direction.y * NEW_DIRECTION_MODIFIER,
-                            position.z + direction.z * NEW_DIRECTION_MODIFIER
-                    );
-                    Vector3f impulseVector = new Vector3f(direction);
-                    entity.send(new DropItemRequest(heldItem, entity, impulseVector, newPosition));
+            if (entity.hasComponent(PlayerCharacterComponent.class)) {
+                blackFlagSlot = searchInventoryForFlag(entity, LASUtils.BLACK_FLAG_URI);
+                if (blackFlagSlot != EntityRef.NULL) {
+                    Vector3f position = new Vector3f(entity.getComponent(LocationComponent.class).getLocalPosition());
+                    Vector3f impulseVector = new Vector3f(attackingPlayer.getComponent(LocationComponent.class).getLocalPosition());
+                    entity.send(new DropItemRequest(blackFlagSlot, entity, impulseVector, position));
+                    return;
+                }
+                redFlagSlot = searchInventoryForFlag(entity, LASUtils.RED_FLAG_URI);
+                if (redFlagSlot != EntityRef.NULL) {
+                    Vector3f position = new Vector3f(entity.getComponent(LocationComponent.class).getLocalPosition());
+                    Vector3f impulseVector = new Vector3f(attackingPlayer.getComponent(LocationComponent.class).getLocalPosition());
+                    entity.send(new DropItemRequest(redFlagSlot, entity, impulseVector, position));
+                    return;
+                }
+            }
+        }
+    }
+
+    // Checks to see if player picks up flag of the same team
+    // If so, moves flag back to base
+    @ReceiveEvent
+    public void onInventorySlotChanged(InventorySlotChangedEvent event, EntityRef entity) {
+        EntityRef playerEntity = entity;
```

Comment:
I think you could simply have the method define the EntityRef as `playerEntity` initially, but it'd be better to create a separate methods for this too-

0. Create a method that has all of the logic here, call it- `moveFlagToBaseIfPickedBySameTeamPlayer()` (or a better shorter version if you come up with any.
1. `private String checkWhichFlagPicked(event)`: This gets the new picked item, checks if it has BlockItemComponent, and if it does,
    - it checks if it has the RedFlagComponent (returns `LASUtils.RED_TEAM`)
    - it checks if it has the BlackFlagComponent (returns `LASUtils.BLACK_TEAM`)
    - else (returns `LASUtils.RED_TEAM`)
2. `private boolean isPlayerInTeam(String team)`: Check if player is on the same team as returned team from above method.
3. Inside LASUtils create a method `private Vector3i getFlagLocation(String team)` that returns the flag location as per the team passed.

Everything combined would look like this inside `moveFlagToBaseIfPickedBySameTeamPlayer`
```
String flagTeam = checkWhichFlagPicked(event);
if (isPlayerInTeam(flagTeam) {
    worldProvider.setBlock(LASUtils.getFlagLocation(team), blockManager.getBlock(LASUtils.RED_FLAG_URI))
    inventoryManager.removeItem(playerEntity, EntityRef.NULL, item, true, 1);
}
```